File: internal/arr/radarr.go
package arr

import (
	"context"
	"time"

	"github.com/zekurio/arr-cal-proxy/internal/config"
	"github.com/zekurio/arr-cal-proxy/internal/event"
)

type radarrMovie struct {
	ID              int64      `json:"id"`
	Title           string     `json:"title"`
	InCinemas       time.Time  `json:"inCinemas"`
	DigitalRelease  time.Time  `json:"digitalRelease"`
	PhysicalRelease time.Time  `json:"physicalRelease"`
	HasFile         bool       `json:"hasFile"`
	Status          string     `json:"status"`
	Overview        string     `json:"overview"`
	Images          []arrImage `json:"images"`
}

type movieRelease struct {
	name string // UID fragment
	kind event.Kind
	date time.Time
}

func fetchRadarr(ctx context.Context, inst config.Instance, start, end time.Time) ([]event.Event, error) {
	q := calendarQuery(start, end, inst.IncludeUnmonitored)

	var movies []radarrMovie
	if err := getJSON(ctx, inst, "/api/v3/calendar", q, &movies); err != nil {
		return nil, err
	}

	var events []event.Event
	for _, m := range movies {
		releases := []movieRelease{
			{"cinema", event.KindMovieCinema, m.InCinemas},
			{"digital", event.KindMovieDigital, m.DigitalRelease},
			{"physical", event.KindMoviePhysical, m.PhysicalRelease},
		}
		for _, rel := range releases {
			if rel.date.IsZero() {
				continue
			}
			// Release timestamps are calendar dates, not instants: keep the
			// date as given (UTC) instead of converting through local zones.
			day := time.Date(rel.date.Year(), rel.date.Month(), rel.date.Day(), 0, 0, 0, 0, time.UTC)
			// Radarr returns the movie if any of its dates is in the window;
			// filter each release date individually.
			if day.Before(start) || !day.Before(end) {
				continue
			}
			events = append(events, event.Event{
				UID:        event.MovieUID(inst.Name, m.ID, rel.name),
				Instance:   inst.Name,
				Source:     config.TypeRadarr,
				Kind:       rel.kind,
				Title:      m.Title,
				Start:      day,
				End:        day.AddDate(0, 0, 1),
				AllDay:     true,
				Downloaded: m.HasFile,
				Overview:   m.Overview,
				PosterURL:  posterURL(m.Images),
			})
		}
	}
	return events, nil
}

File: src/upstream/radarr.ts
import { Type } from '@sinclair/typebox'

import type { Instance } from '../config.ts'
import { addUtcDays, type CalendarEvent, movieUid } from '../domain/event.ts'
import { calendarQuery, getJson, type HttpFetch, posterUrl } from './client.ts'
import { ArrImageSchema } from './schemas.ts'

const NullableStringSchema = Type.Union([Type.String(), Type.Null()])
const RadarrMovieSchema = Type.Object({
  id: Type.Number(),
  title: Type.String(),
  tmdbId: Type.Optional(Type.Number()),
  inCinemas: Type.Optional(NullableStringSchema),
  digitalRelease: Type.Optional(NullableStringSchema),
  physicalRelease: Type.Optional(NullableStringSchema),
  hasFile: Type.Optional(Type.Boolean()),
  overview: Type.Optional(NullableStringSchema),
  images: Type.Optional(Type.Union([Type.Array(ArrImageSchema), Type.Null()])),
})
const RadarrCalendarSchema = Type.Array(RadarrMovieSchema)

function releaseDay(value: string | null | undefined): Date | undefined {
  if (!value) {
    return undefined
  }
  const date = new Date(value)
  if (Number.isNaN(date.getTime())) {
    return undefined
  }
  return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()))
}

export async function fetchRadarr(
  instance: Instance,
  start: Date,
  end: Date,
  fetchFn: HttpFetch = fetch,
): Promise<CalendarEvent[]> {
  const query = calendarQuery(start, end, instance.includeUnmonitored)
  const movies = await getJson(instance, '/api/v3/calendar', query, RadarrCalendarSchema, fetchFn)
  const events: CalendarEvent[] = []

  for (const movie of movies) {
    const releases = [
      { name: 'cinema' as const, kind: 'movie-cinema' as const, date: movie.inCinemas },
      { name: 'digital' as const, kind: 'movie-digital' as const, date: movie.digitalRelease },
      { name: 'physical' as const, kind: 'movie-physical' as const, date: movie.physicalRelease },
    ]

    for (const release of releases) {
      const day = releaseDay(release.date)
      if (day === undefined || day < start || day >= end) {
        continue
      }
      events.push({
        uid: movieUid(instance.name, movie.id, release.name),
        instance: instance.name,
        source: 'radarr',
        kind: release.kind,
        title: movie.title,
        subtitle: '',
        season: 0,
        episode: 0,
        start: day,
        end: addUtcDays(day, 1),
        allDay: true,
        downloaded: movie.hasFile ?? false,
        overview: movie.overview ?? '',
        posterUrl: posterUrl(movie.images),
        providerIds: movie.tmdbId ? { Tmdb: String(movie.tmdbId) } : {},
        jellyfinUrl: '',
      })
    }
  }

  return events
}
